src/components/utils/src/conditional_variable_boost.cc

```diff
@@ -62,22 +63,26 @@ void ConditionalVariable::Broadcast() {
 bool ConditionalVariable::Wait(BaseLock& lock) {
   // NOTE this grossness is due to boost mutex and recursive mutex not sharing a
   // superclass
-
-  lock.AssertTakenAndMarkFree();
-  // What kind of lock are we ?
-  if (Lock* test_lock = dynamic_cast<Lock*>(&lock)) {
-    // Regular lock
-    cond_var_.wait<boost::mutex>(test_lock->mutex_);
-  } else if (RecursiveLock* test_rec_lock =
-                 dynamic_cast<RecursiveLock*>(&lock)) {
-    // Recursive lock
-    cond_var_.wait<boost::recursive_mutex>(test_rec_lock->mutex_);
-  } else {
-    // unknown, give up the lock
-    LOG4CXX_ERROR(logger_, "Unknown lock type!");
-    NOTREACHED();
+  try {
+    lock.AssertTakenAndMarkFree();
+    // What kind of lock are we ?
+    if (Lock* test_lock = dynamic_cast<Lock*>(&lock)) {
+      // Regular lock
+      cond_var_.wait<boost::mutex>(test_lock->mutex_);
+    } else if (RecursiveLock* test_rec_lock =
+                   dynamic_cast<RecursiveLock*>(&lock)) {
+      // Recursive lock
+      cond_var_.wait<boost::recursive_mutex>(test_rec_lock->mutex_);
+    } else {
+      // unknown, give up the lock
+      LOG4CXX_ERROR(logger_, "Unknown lock type!");
+      NOTREACHED();
+    }
+    lock.AssertFreeAndMarkTaken();
+  } catch (const boost::exception& error) {
+    std::string error_string = boost::diagnostic_information(error);
+    LOG4CXX_FATAL(logger_, error_string);
```

Comment:
I was under the assumption that `LOG4CXX_FATAL` caused the programs to close, but apparently not. I think a `NOTREACHED()` should be added here